lib/Sema/CSDiag.cpp

```diff
@@ -5980,18 +5987,69 @@ bool FailureDiagnosis::diagnoseArgumentGenericRequirements(
     return false;
 
   class RequirementsListener : public GenericRequirementsCheckListener {
+    ConstraintSystem &CS;
+    AbstractFunctionDecl *Candidate;
+
+    Expr *CallExpr;
+    Expr *FnExpr;
+    Expr *ArgExpr;
+
   public:
+    RequirementsListener(ConstraintSystem &cs, AbstractFunctionDecl *AFD,
+                         Expr *callExpr, Expr *fnExpr, Expr *argExpr)
+        : CS(cs), Candidate(AFD), CallExpr(callExpr), FnExpr(fnExpr),
+          ArgExpr(argExpr) {}
+
     bool shouldCheck(RequirementKind kind, Type first, Type second) override {
       // This means that we have encountered requirement which references
       // generic parameter not used in the arguments, we can't diagnose it here.
       return !(first->hasTypeParameter() || first->isTypeVariableOrMember());
     }
+
+    void unsatisfiedRequirement(const Requirement &req, Type first,
+                                Type second) override {
+      switch (req.getKind()) {
+      case RequirementKind::Conformance:
+      case RequirementKind::Layout:
+        return;
```

Comment:
I bet you could customize the `Conformance` case here to improve diagnostics when one of the type arguments (or associated types thereof) fails to meet some conformance requirement.